lib/buckets/common.js

```diff
@@ -151,3 +250,494 @@ module.exports = {
     listBuckets: listBuckets,
     listObjects: listObjects
 };
+
+///--- Internal
+
+/*
+ * LimitMarkerStream
+ *
+ * This class provides a seamless wrapper on top of an object stream that
+ * requires a limit and optionally a marker.  LimitMarkerStream will create the
+ * stream automatically under-the-hood when needed, and will recreate the
+ * stream with a new marker when the limit has been hit (automatic pagination).
+ *
+ * The constructor takes a single oject with the following options:
+ *
+ * log        Bunyan   [required] Bunyan logger object.
+ * limit      Number   [required] The limit to use.
+ * markerKey  String   [required] The object key which contains the marker.
+ * marker     String   [optional] Initial marker to use.
+ * getStream  Function [required] Function to retrieve a new stream, will be
+ *                                called with the following signature:
+ *                                opts.getStream(currentMarker, currentLimit)
+ *                                and must return a new Stream to use.
+ *
+ * With an LimitMarkerStream like:
+ *
+ *  var lmstream = new LimitMarkerStream(opts);
+ *
+ * The `getNextRecord` method can be used:
+ *
+ *  lmstream.getNextRecord(record, isDone) {
+ *    if (isDone) {
+ *      return;
+ *    }
+ *    // record -> a single object from the underlying stream.read()
+ *  });
+ *
+ * `getNextRecord` will handle automatically paginating the stream when needed,
+ * without the user needing to be concerned with any of that logic.
+ *
+ * The `setNewMarker` method can be used to update the marker by throwing out
+ * all records until the new marker is seen.  It is an error to pass a Marker
+ * to this that is less-than (lexicographical sort) the current marker.
+ *
+ * Finally, the `lmstream` object has a `done` property set when the stream has
+ * fully been exhausted.
+ */
+util.inherits(LimitMarkerStream, EventEmitter);
+function LimitMarkerStream(opts) {
```

Comment:
This looks different to the node-lomstream library we have, but do they serve the same purpose?

Reply:
So, this was originally a LimitOffset (no marker) stream before the listing API was updated a couple of months ago.  I had already written code for that, so I just adapted it to work with a Limit and a Marker.

It probably could be rewritten to use node-lomstream at this point (though I have little experience with that library).  I don't know if they do exactly the same thing however.

Reply:
> I don't know if they do exactly the same thing however.

Yea me neither off hand. Seems fine to keep muskie/buckets-api's own implementation if it's already working!